mypy/test/data.py

```diff
@@ -347,12 +347,14 @@ def reportinfo(self) -> tuple[str, int, str]:
         return self.file, self.line, self.name
 
     def repr_failure(self, excinfo: Any, style: Any | None = None) -> str:
-        if excinfo.errisinstance(SystemExit):
+        if isinstance(excinfo.value, SystemExit):
             # We assume that before doing exit() (which raises SystemExit) we've printed
             # enough context about what happened so that a stack trace is not useful.
             # In particular, uncaught exceptions during semantic analysis or type checking
             # call exit() and they already print out a stack trace.
             excrepr = excinfo.exconly()
+        elif isinstance(excinfo.value, pytest.fail.Exception) and not excinfo.value.pytrace:
+            excrepr = excinfo.exconly()
```

Comment:
pytest handles the `pytrace` attribute when it's producing a traceback
https://github.com/pytest-dev/pytest/blob/5dbfb8e108e04d287a47527aa1ede8d6371f8b0b/src/_pytest/nodes.py#L447-L449
but since we do our own traceback here (for whatever reason), we should handle it on our own